contracts/strategy/TradeFintechStrategy.sol

```diff
@@ -0,0 +1,150 @@
+        totalInvested += _amount;
+    }
+
+    function _freeFunds(uint256 _amount) internal override {
+        require(block.timestamp > lockPeriodEnds, "Lock period has not ended");
+        // We don't check available liquidity because we need the tx to
+        // revert if there is not enough liquidity so we don't improperly
+        // pass a loss on to the user withdrawing.
+        asset.safeTransferFrom(TokenizedStrategy.management(), address(this), _amount);
+        totalInvested -= _amount;
+    }
+
+    /**
+     * @dev Optional function for a strategist to override that will
+     * allow management to manually withdraw deployed funds from the
+     * yield source if a strategy is shutdown.
+     *
+     * This should attempt to free `_amount`, noting that `_amount` may
+     * be more than is currently deployed.
+     *
+     * NOTE: This will not realize any profits or losses. A separate
+     * {report} will be needed in order to record any profit/loss. If
+     * a report may need to be called after a shutdown it is important
+     * to check if the strategy is shutdown during {_harvestAndReport}
+     * so that it does not simply re-deploy all funds that had been freed.
+     *
+     * EX:
+     *   if(freeAsset > 0 && !TokenizedStrategy.isShutdown()) {
+     *       depositFunds...
+     *    }
+     *
+     * @param _amount The amount of asset to attempt to free.
+     */
+    function _emergencyWithdraw(uint256 _amount) internal override {
+        uint256 amountToWithdraw = Math.min(_amount, asset.balanceOf(address(this)));
+        asset.transfer(TokenizedStrategy.management(), amountToWithdraw);
+        totalInvested = totalInvested > amountToWithdraw ? totalInvested - amountToWithdraw : 0;
+    }
+
+    /// @notice Allows the manager to report gains or losses.
+    /// @dev Should be called before calling report() to report the amount of the gain or loss.
+    /// @dev The manager can only report gains or losses.
+    /// @param _gain The amount of the gain.
+    /// @param _loss The amount of the loss.
+    function reportGainOrLoss(uint256 _gain, uint256 _loss) external onlyManagement {
+        if (_gain > 0) {
+            require(_loss == 0, "Cannot report both gain and loss");
+            totalGains += _gain;
+            totalInvested += _gain;
+            emit GainReported(msg.sender, _gain);
+        } else if (_loss > 0) {
+            require(_loss <= totalInvested, "Cannot report loss more than total invested");
+            totalLosses += _loss;
+            totalInvested -= _loss;
+            emit LossReported(msg.sender, _loss);
+        }
+    }
+
+    /// @notice Allows the manager to return the funds.
+    /// @param _amount The amount that is being returned.
+    function returnFunds(uint256 _amount) external onlyManagement {
+        require(_amount > 0, "Amount must be greater than 0.");
+        require(totalInvested > 0, "No funds to return.");
```

Comment:
I believe instead of checking that `totalInvested > 0` it's better to be sure that `_amount >= totalInvested` otherwise we might have a underflow error here: `totalInvested -= _amount;`

Reply:
Actually I think here we should check `_amount <= totalInvested`

This functions is for the Strategy Manager to return funds after making some investments

Reply:
yeah, you are right, I made a typo

Reply:
` require(totalInvested > 0, "No funds to return.");` is not needed anymore `require(_amount <= totalInvested` check will handle 0 value for `totalInvested`